### ISSUE TITLE
PayPalLifecycleObserver lives forever / can not be removed

### ISSUE THREAD

**Issue by StefMa** [braintree/braintree_android#554]
### General information

* SDK/Library version: 4.10.1
* Environment: Prod and Sandbox
* Android Version and Device: Doesn't matter
* Braintree dependencies:
```
deps.braintreePaypal = "com.braintreepayments.api:paypal:4.10.1"
deps.braintreeDataCollector = "com.braintreepayments.api:data-collector:4.10.1"
```

### Issue description

We have the following flow of adding an PayPal Account to our app:
<img src="https://user-images.githubusercontent.com/10229883/169029493-e9be2278-7547-48ad-8385-7f99678f23ec.gif" width=280 />

This is what you see:
* A Fragment with a FloationActionButton
* On click a new Fragment (name it `PayPalFragment`) will be added
* The PaypalFragment creates a `PayPalClient` (somehow) and sets the `PayPalListener` to it.
* Then we are calling `payPalClient.tokenizePayPalAccount(activity, request)` 
* This call leads to opening the Browser on the PapPal website
* The user closes the Browser and land again on the `PayPalFragment`
* The `PayPalFragment` recognize the "abbording" and ask if the user wants to "Retry"
* On `Retry` we're doing the same as above (create a new `PayPalClient`, set the `Listener` and call `tokenizePayPalAccount`)

Then the gifs ends 🙃 

Anyways, what happen if the user closes the Browser again? 
Or even complete the flow?
Our app will crash! 

The reason is:
We use Kotlin Coroutines and wrap the Listener into a `suspendingCoroutine`:
```kotlin
fun startPayPalFlow() = suspendCoroutine<String> {
    val listener = Listener(continuation)
    payPalClient.setListener(listener)
    payPalClient.tokenizePayPalAccount(activity, request)
}

class Listener(private val continuation: Continuation<String>) : PayPalListener {
    override fun onPayPalSuccess(payPalAccountNonce: PayPalAccountNonce) {
        continuation.resume(payPalAccountNonce.string)
    }

    override fun onPayPalFailure(error: java.lang.Exception) {
        continuation.resumeWithException(error)
    }

}
```

The app crashes here with a `IllegateStateException: Already resumed`.
Which is intented, in case `resume[WithException]` will be called more than one time.
[(See also Kotlin docs here)](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.coroutines/suspend-coroutine.html)

So what happen behinde the scenes?
As soon as we create a `PayPalClient` a `PayPalLifecycleObserver` will be created:
https://github.com/braintree/braintree_android/blob/21b38715924d63774034fd7167ceb2ec3e362264/PayPal/src/main/java/com/braintreepayments/api/PayPalClient.java#L64-L71

This Observer gets the Lifecycle of the current Fragment (or Activity).
This means, each time we create *a new client*, a *new Observer* will be created.

The Observer does, `onResume`, calling on the given `PayPalClient` the `onBrowserSwitchResult` method, which will call the added Listener.
https://github.com/braintree/braintree_android/blob/21b38715924d63774034fd7167ceb2ec3e362264/PayPal/src/main/java/com/braintreepayments/api/PayPalLifeCycleObserver.java#L60-L62
https://github.com/braintree/braintree_android/blob/21b38715924d63774034fd7167ceb2ec3e362264/PayPal/src/main/java/com/braintreepayments/api/PayPalClient.java#L278-L280

The problem is now (and this is why our app crashes):
Each time `onResume` on the given `Fragment` (or Activity) will be called, the Listener will be called.
This means, that the "first" `PayPalClient` and the "first" `PayPalListener` will be called again, in case we "Retry".
As the Fragment-Lifecycle is still active. No Observers are removed. But a new `PayPalClient` was created.

I guess it make sense if you add an API to **either** remove the Listener, remove the Observer or "clean" the `PayPalClient` (or something).

tl;dr:
Each `PayPalClient` creates a new `PayPalLifecycleObserver` and bound it to the Fragment/Activity.
In case you create multiple PayPalClients (for whatever reasons) you create multiple `PayPalLifecycleObserver` which are here as long as the Fragment/Activity is alive.
Imagine you have a Single Activity Architecture and put the `Activity` inside, the Observer lifes forever and even leak the `PayPalClient` which might never be reused within the app.

**Comment by StefMa** [comment]
Hi @sshropshire,

thanks for the fast response!

> * Can a single PayPalClient instance be used here to prevent multiple observers from being created?

Unfourtnaly not with our current setup.
As the `PayPalClient` requires a `BraintreeClient`. The `BraintreeClient` can only be constructed with an `token`. But the `token` has to be fetched for several reasons from our backend. It is not hardcoded in the app.
So it is quite hart with a nice architecture to do:
* Fetch the token in the `domain` layer
* Going back to the `UI` layer to create a `PayPalClient` (because it also requires an Activity or an Fragment)
* Put the PayPalClient back to the `domain` layer to do some work

A solution would be to do all the work from the PayPalClient on the `presentation` layer but this is not a good architecture for us at least.

> * Does this issue still occur without wrapping the listener in a coroutine?

In this case (and when using the following code*1)

*1:
```kotlin
```

**Comment by StefMa** [comment]
Hi @sshropshire,

thanks for the fast response!

> * Can a single PayPalClient instance be used here to prevent multiple observers from being created?

Unfourtnaly not with our current setup.
As the `PayPalClient` requires a `BraintreeClient`. The `BraintreeClient` can only be constructed with an `token`. But the `token` has to be fetched for several reasons from our backend. It is not hardcoded in the app.
So it is quite hart with a nice architecture to do:
* Fetch the token in the `domain` layer
* Going back to the `UI` layer to create a `PayPalClient` (because it also requires an Activity or an Fragment)
* Put the PayPalClient back to the `domain` layer to do some work

A solution would be to do all the work from the PayPalClient on the `presentation` layer but this is not a good architecture for us at least.

> * Does this issue still occur without wrapping the listener in a coroutine?

In this case (and when using the following code *1 for the listener) it is at least not crashing, but the **second and following** PayPal flows are not working anymore 🙈 . (I guess coroutine or not has nothing to do with the underlining issue).

The reason why it is not working anymore is because the "second (or following)" Listeners will **not be called** by the `Observer`. Only the **very first** Listener will be called. This is because of the `if` condition in the Observer:
https://github.com/braintree/braintree_android/blob/21b38715924d63774034fd7167ceb2ec3e362264/PayPal/src/main/java/com/braintreepayments/api/PayPalLifeCycleObserver.java#L59-L60

The second (or following) Observers get here an `null` `BrowserSwitchResult` returned from the `payPalClient.getBrowserSwitchResult`. I'm not sure why this is like that, but this is the case.

*1:
```kotlin
class Listener(private val continuation: CancellableContinuation<String>) : PayPalListener {
    override fun onPayPalSuccess(payPalAccountNonce: PayPalAccountNonce) {
        if (continuation.isActive) continuation.resume(payPalAccountNonce.string)
    }

    override fun onPayPalFailure(error: java.lang.Exception) {
        if (continuation.isActive) continuation.resumeWithException(error)
    }

}
```

Looking forward for a release which includes a `destroy()` method 🤞

**Comment by sshropshire** [comment]
Hi @StefMa thanks for your patience. We're working on prioritization at the moment. We may enable SNAPSHOT builds soon so you can test out functionality before we do an official release.

Also, have you looked at our [ClientTokenProvider](https://github.com/braintree/braintree_android/blob/master/v4.9.0%2B_MIGRATION_GUIDE.md#client-token-provider) constructor? It allows merchant apps a way to define a client token fetching strategy for the SDK to use when it needs a client token.

**Comment by StefMa** [comment]
We postponed the update and update to library `4.23.1` today.
It works now.
I have no idea why 🤷 
Maybe the changes here were the fix 🤷 
https://github.com/braintree/braintree_android/blob/29b64ca0a41c5894d4e61184573b2bf41508a984/PayPal/src/main/java/com/braintreepayments/api/PayPalLifeCycleObserver.java#L60-L76

**Comment by sshropshire** [comment]
Thanks @StefMa ! We did make an update to our internal Browser Switching logic to guard against launching a Chrome Custom Tab when the host Activity is finishing. That may have helped in this case.

Feel free to re-open if the issue comes back up again.
